### ISSUE TITLE
Support MicroProfile 4.0

### ISSUE THREAD

**Issue by Emily-Jiang** [OpenLiberty/open-liberty#11014]
We need to implement MicroProfile 4.0 feature to be released in 4Q 2020. Click [here ](https://github.com/eclipse/microprofile/milestone/5) for more info.
This release will rebase Jakarta EE8. I assume all the sub specs need to be a update to consume the new maven coordinates plus some potential backward incompatible changes.

- Open API 2.0
- Rest Client 2.0
- Metrics 3.0
- Config 2.0
- Fault Tolerance 3.0
- JWT 1.2
- Health 3.0
- Open Tracing 2.0

When ready, add links to the Upcoming Feature Overview document as well as Feature Test Summary and blog post issues:
- UFO: https://ibm.ent.box.com/file/711400637013?s=2tqwnonrkl18fo4l21xlkcjnptddm5xx
- FTS: https://github.com/OpenLiberty/open-liberty/issues/15253
- Beta Blog Post (if applicable):
- Blog Post: 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([Adam Yoho](https://github.com/ayoho), [Martin Holder](https://github.com/mhldr), or [Dave Waddling](https://github.com/dave-waddling)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Kareen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [ ] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by cthigh** [comment]
UFO review on July 21, 2020.  

Meeting notes:
1. UFO link not in the issue.
2. Page 7 - Question - how does the existing REST metric cause a skew?  (3rd bullet - “The existing REST metric ,REST.request which tracks count and cumulative elapsed time duration, can have its data skewed by REST requests that end with an unmapped exception”)

 	Has this gone far enough?  Useful to have a count of how many completed with a particular exception?
	Action - continue with discussions for future versions of spec - break out different exceptions for REST metrics. 
 
3. Emily brought up discussion on metrics going into different namespaces.  When REST endpoints are annotated in the application - it will be in the application namespace?  REST metrics will be in the base namespace?  Grafana has access to all. 
        Adam Bean (sp). Get user feedback on namespaces - debate on whether there is too much segmenting or not enough.

4. Page 15 - Don surprised that mapped exception in a tag rather than part of name.  Revisit offline.  Action item. 
5. Page 19 - typo  “produces”
6. P 23 - Emily question to follow up on. 
7. At end Emily brought up dependencies - Will depend on new mpConfig feature.  No backward compatibility. 
8. Jared - is there a change to use Small Rye with mpMetrics-3.0?    don’t think so.  Considering after 3.0
9. Migration - Alasdair asked about the migration tools - In the past we have not added microprofile to the migration tools - Previously we brought up adding differences be added to MicroProfile LSP.  Revisit with Emily in context of the umbrella feature.  https://github.com/OpenLiberty/open-liberty/issues/11014

**Comment by cthigh** [comment]
At the MP metrics 3.0 UFO review (July 21), Alasdair asked that the migration topic should be handled in this umbrella feature level to decide how migration differences should be handled from a tools point of view.

**Comment by mbroz2** [comment]
UFO Review Comments:
- (End User Overview Slide) We will not be updating the JEE feature versions to Jakarta EE 8 in order to release MP 4.0.
- (Not specific to Liberty) Verify whether MP JWT needs to be updated to make sure it's not introducing a breaking change (provided scope in pom)
- (Communication Slide) Open an issue for ol.io guides (guides-common repo) to track necessary work.
- (Feature Design slide) Prevent MP 2.x, 3.x JEE features from being configured with MP 4.0 (not supporting mixing)
- (Not Liberty specific) Investigate providing an uberjar for MP 4.0
- (Developer Experience slide): Add LSP
- (Platform/Cloud Considerations slide): Be explicit about Java 'SE' where applicable (done). Add latest supported version of Java SE (14 now, possibly 15 by release time).
- (Platform/Cloud Considerations slide): POC Agenda item for removing packages/zips of previous release versions
- (Migration Impact - Config slide) Single box note/folder that keeps track of all migration impacts introduced by MP (Expectation is that MP provides a single place for incompatible changes and we just need a link to it).

**Comment by Emily-Jiang** [comment]
All comments on UFO have been addresses and the UFO can be found [here](https://ibm.box.com/s/2tqwnonrkl18fo4l21xlkcjnptddm5xx ).

**Comment by Emily-Jiang** [comment]
@steven1046 can you please add the accessibility approval for this feature as this feature does not introduce any GUI?

**Comment by Emily-Jiang** [comment]
@samwatibm This is ready for globalization approval. This feature is a umbrella feature for MicroProfile 4.0 and it does not introduce any error messages except feature description text.

**Comment by samwatibm** [comment]
@Emily-Jiang Can you confirm the feature description text either doesn't need to be translated or is already translated?   Translations for 21.0.0.1 are not yet back.

**Comment by Emily-Jiang** [comment]
yes, as I mentioned as above that there is no translation for this feature.

**Comment by Emily-Jiang** [comment]
@chirp1  this feature is umbrella release and it does not add any more functionality, so it is n/a for ID. Can you add ID approval for this feature?

**Comment by Emily-Jiang** [comment]
@skasund can you please give STE approval as this is no op for STE because MP 4.0 does not deliver any functionality and it is just an aggregation of other MP specs included by MP 4.0.

**Comment by skasund** [comment]
@Emily-Jiang   I've just provided the STE approval

**Comment by chirp1** [comment]
No ID work required for this epic. Approving.
